common/src/Renderer/EntityDecalRenderer.cpp

```diff
@@ -0,0 +1,484 @@
+  // also do nothing if the VBO storage fields are null, but it shouldn't happen
+  if (!data.texture || data.vertexHolderKey == nullptr || data.faceIndicesKey == nullptr)
+  {
+    data.validatedGeometry = false;
+    return;
+  }
+
+  // update the VBO
+  m_vertexArray->deleteVerticesWithKey(data.vertexHolderKey);
+
+  const auto faceIndexHolder = m_faces->at(data.texture);
+  faceIndexHolder->zeroElementsWithKey(data.faceIndicesKey);
+
+  if (!faceIndexHolder->hasValidIndices())
+  {
+    // there are no indices left to render for this texture
+    m_faces->erase(data.texture);
+  }
+
+  data.vertexHolderKey = nullptr;
+  data.faceIndicesKey = nullptr;
+  data.validatedGeometry = false;
+}
+
+void EntityDecalRenderer::validateDecalBrushes(
+  const Model::EntityNode* entityNode, EntityDecalData& data) const
+{
+  if (data.validatedBrushes)
+  {
+    // already validated, no need to check again
+    return;
+  }
+
+  const auto document = kdl::mem_lock(m_document);
+  const auto world = document->world();
+
+  // create a fake brush to represent the entity for intersection purposes
+  // todo: (?) i only do this because collectTouchingNodes only accepts a brush
+  // and i can't find a bbox version of this method
+  const auto bounds = entityNode->physicalBounds();
+  const auto builder = Model::BrushBuilder(
+    world->mapFormat(), document->worldBounds(), document->game()->defaultFaceAttribs());
+  auto brush = builder.createCuboid(bounds, "").value();
+  auto brushNode = Model::BrushNode(brush);
+
+  // collect all the brush nodes that touch the fake brush
+  const auto touching = Model::collectTouchingNodes(
+    std::vector<Model::Node*>{world}, std::vector{&brushNode});
+
+  // track them in the entity
+  data.brushes.clear();
+  for (auto node : touching)
+  {
+    if (const auto* br = dynamic_cast<Model::BrushNode*>(node))
+    {
+      // ignore hidden brushes
+      if (br->visible())
+      {
+        data.brushes.insert(br);
+      }
+    }
+  }
+
+  data.validatedBrushes = true;
+}
+
+void EntityDecalRenderer::validateDecalGeometry(
+  const Model::EntityNode* entityNode, EntityDecalData& data) const
+{
+  if (data.validatedGeometry)
+  {
+    // already validated, no need to check again
+    return;
+  }
+
+  const auto spec = getDecalSpecification(entityNode);
+  if (spec == std::nullopt)
+  {
+    // we shouldn't get here
+    data.validatedGeometry = true;
+    return;
+  }
+
+  const auto document = kdl::mem_lock(m_document);
+
+  // shrink the bounds just a tiny bit so we get exclusive checks instead of inclusive
+  const auto bbox = entityNode->physicalBounds().expand(-0.01);
+
+  auto* texture = document->textureManager().texture(spec->textureName);
+  data.texture = texture;
+
+  if (!texture)
+  {
+    // no decal texture was found, don't generate any geometry
+    data.validatedGeometry = true;
+    return;
+  }
+
+  auto vertices = std::vector<Vertex>{};
+  auto indices = std::vector<size_t>{};
```

Comment:
Mostly because I would have liked to upload the data directly into the VBO for each brush, rather than putting them into a list first and then uploading it as a second step. Perhaps I'm overthinking it?

Reply:
I think it's fine. It might even be faster this way. These things are hard to tell without actually measuring.

Reply:
Essentially if the framerate is okay, don't worry about it. If it turns out to be a problem at some point, we can invest time in optimization. Chances are that there are more important places to improve anyway.